Step2: 문자열 계산기

Gilssung opened the pull request:
다음과 같은 기능을 추가하였습니다.

- 사칙연산 계산기 구현
- 입력값에 대한 예외 처리 구현
- 계산기 기본 테스트 구현

리뷰 부탁드리겠습니다

qyu-st committed:
study: 코틀린 기초

qyu-st committed:
study: 코틀린 기초

qyu-st committed:
study: 코틀린 기초

qyu-st committed:
feat: 2단계 - 문자열 계산기

* 사칙연산 계산기 구현
* 입력값에 대한 예외 처리 구현
* 계산기 기본 테스트 구현

galcyurio commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,60 @@
+package step2
+
+class StringCalculator(private val inputStr: String) {
```
계산기가 특정 하나의 계산을 수행할 수 있겠네요.
현실세계의 계산기처럼 여러 번 수행할 수 있도록 변경해보면 어떨까요?

galcyurio commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private fun calculate(accum: Double, operators: List<String>, operands: List<Double>): Double =
+        if (operands.isEmpty())
+            accum
+        else
+            calculate(
+                operate(operators.first(), accum, operands.first()),
+                operators.drop(1),
+                operands.drop(1)
+            )
```
[공식 컨벤션](https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements)에 따르면 if문과 when문은 2줄 이상이면 중괄호를 사용합니다.

galcyurio commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,60 @@
+                operands.drop(1)
+            )
+
+    private fun operate(operator: String, operand1: Double, operand2: Double): Double =
```
Operator라는 별도의 class로 분리해보면 어떨까요?

galcyurio commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,60 @@
+        val operators = operatorPreProcessing()
+
+        require(operands.size == operators.size) {
+            "피연산자의 수(${operands.size})와 연산자 수(${operators.size - 1})가 적합하지 않습니다 " +
+                    "ex. 피연산자 수 == 연산자 수 + 1"
```
디버깅에 필요한 정보를 잘 담고 있는 오류 메시지 👍

galcyurio commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,73 @@
+
+internal class StringCalculatorTest {
+
+    @Test
+    fun `arithmetic operation test`() {
+        val inputStr1 = "2 + 3"
+        val expected1 = 5.0
+        assertEquals(expected1, StringCalculator(inputStr1).result())
+
+        val inputStr2 = "8 - 1"
+        val expected2 = 7.0
+        assertEquals(expected2, StringCalculator(inputStr2).result())
+
+        val inputStr3 = "3 * 4"
+        val expected3 = 12.0
+        assertEquals(expected3, StringCalculator(inputStr3).result())
+
+        val inputStr4 = "8 / 4"
+        val expected4 = 2.0
+        assertEquals(expected4, StringCalculator(inputStr4).result())
+    }
```
JUnit5의 Parameterized Test를 활용해보세요.
https://www.baeldung.com/parameterized-tests-junit-5

galcyurio reviewed (changes requested):
2단계 미션 고생하셨습니다.
코멘트를 남겨두었으니 확인 후 다시 요청해주세요.

qyu-st committed:
refactor: if문 중괄호 사용

공식 컨벤션에 따라 if문이 2줄 이상일 때 중괄호 사용

qyu-st committed:
refactor: Parameterized Test로 전환

JUnit5의 Parameterized Test를 활용해보세요.
https://www.baeldung.com/parameterized-tests-junit-5

qyu-st committed:
feat: ExpressionParser 구현

Input String을 파싱하는 로직을 클래스로 따로 구현

qyu-st committed:
feat: ExpressionParser 단위테스트 구현

qyu-st committed:
feat: Operator 클래스 구현

galcyurio commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -1,73 +1,48 @@
+        "8 / 4,2.0"
+    )
+    fun `arithmetic operation test`(inputStr: String, expected: Double) {
+        assertEquals(expected, StringCalculator(inputStr).result())
```
assertj의 `assertThat()`을 활용해보세요.
훨씬 더 다양한 방법으로 assertion 할 수 있습니다.

galcyurio commented on src/main/kotlin/step2/ExpressionParser.kt:
```diff
@@ -0,0 +1,22 @@
+package step2
+
+class ExpressionParser(private val inputStr: String) {
```
한 가지 수식에 대해 operands와 operators를 반환하므로 ExpressionParser라는 네이밍보다는 Expression이 더 적절하지 않을까요?

위와 같은 관점에서 getOperands(), getOperators()와 같은 함수보다는 `expression.operands`, `expression.operators`와 같이 호출할 수 있도록 필드로 선언해보면 좋겠습니다.

galcyurio commented on src/main/kotlin/step2/ExpressionParser.kt:
```diff
@@ -0,0 +1,22 @@
+package step2
```
각 피드백별로 잘 나누어진 커밋 👍

galcyurio commented on src/test/kotlin/step2/ExpressionParserTest.kt:
```diff
@@ -0,0 +1,45 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = ["123 + 삼백", "123 + 3849 ^ 13"])
+    fun `non-arithmetic operator or non-digit check test`(inputStr: String) {
```
테스트 코드에서는 가독성이 굉장히 중요합니다.
함께 일하는 동료중에 외국인이 있지 않은 이상 한글로 쓰는게 본인에게도 좋고 협업하는 팀원들 입장에서도 읽기 좋지 않을까요?

galcyurio commented on src/main/kotlin/step2/ExpressionParser.kt:
```diff
@@ -15,8 +15,26 @@ class ExpressionParser(private val inputStr: String) {
+        }
+    }
+
+    private fun find(targetSymbol: String): Operator {
+        return Operator.values()
+            .find { op ->
+                op.symbol == targetSymbol
+            } ?: run {
+            throw IllegalArgumentException("사칙연산 기호가 아닌 입력이 존재합니다. symbol = $targetSymbol")
+        }
+    }
```
특정 symbol을 찾아서 알맞은 Operator를 생성하는 것이므로 Operator의 역할에 가깝습니다.
Operator의 companion object 내부에 팩토리 함수로 구현해보면 어떨까요?

galcyurio commented on src/main/kotlin/step2/ExpressionParser.kt:
```diff
@@ -15,8 +15,26 @@ class ExpressionParser(private val inputStr: String) {
+        val operators = getOperators()
+            .drop(1)
+            .map { it.symbol }
+        return getOperands().foldIndexed("") { idx, accum, operand ->
```
정해진 건 아니지만 보통 accumulator는 줄여서 `acc`로 이름을 지어줍니다.

https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/common/src/generated/_Sequences.kt#L1261

galcyurio commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,26 @@
+import java.util.function.BinaryOperator
+import java.util.function.DoubleBinaryOperator
+
+enum class Operator(val symbol: String): BinaryOperator<Double>, DoubleBinaryOperator {
```
ktlint가 제대로 적용되지 않은 것 같습니다.

galcyurio reviewed (changes requested):
피드백 잘 반영해주셨습니다 👍 
반영 후 추가된 코드에 대해 코멘트를 남겨두었으니 피드백 반영 후 다시 요청해주세요.

qyu-st committed:
refactor: assertj assertThat 적용

qyu-st committed:
style: ktlint 관련 수정

* Operator.kt의 경우 Gradle task를 통해 수행해도 적용되지 않음
* option + command + l 을 통해서 수정

qyu-st committed:
refactor: ExpressionParser 수정

* 네이밍 수정(Expression, acc)
* 메서드를 필드 호출로 수정(Expression::operands, Expression::operators)

qyu-st committed:
refactor: 함수의 역할에 맞는 클래스로 이동

* Expression::find -> Operator::find

Gilssung committed:
Merge branch 'gilssung' into step2

galcyurio commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -78,7 +66,7 @@ internal class StringCalculatorTest {
             .append(inputStr2)
 
-        assertEquals(expected, calculator.display())
+        assertThat(expected).isEqualTo(calculator.display())
```
actual, expected 매개변수를 반대로 넣으셨습니다.
`assertThat(actual).isEqualTo(expected)`와 같이 작성해주셔야 테스트가 실패했을 때의 에러 메시지가 헷갈리지 않습니다.

galcyurio commented on src/main/kotlin/step2/Expression.kt:
```diff
@@ -1,32 +1,34 @@
     private val validArgRegex = Regex("[\\d */+-]")
     private val operatorRegex = Regex("[+*/-]")
 
+    var operands: List<Double>
+    var operators: List<Operator>
```
`val`로 선언하셔도 됩니다.

galcyurio reviewed (approved):
2단계 미션 고생하셨습니다. 👏 
추가로 몇 가지 코멘트 남겨두었으니 확인해주세요.
이제 본격적으로 자동차 경주 구현하러 가시죠!
